crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -687,16 +673,53 @@ bitflags! {
     }
 }
 
-#[derive(PartialEq, Copy, Clone)]
+#[derive(Debug, PartialEq, Copy, Clone)]
 enum FunctionKind {
+    /// A lambda expression, e.g., `lambda x: x`
     Lambda,
+    /// A function definition, e.g., `def f(x): ...`
     FunctionDef,
 }
 
+impl FunctionKind {
+    /// Returns the token that terminates a list of parameters.
+    const fn list_terminator(self) -> TokenKind {
+        match self {
+            FunctionKind::Lambda => TokenKind::Colon,
+            FunctionKind::FunctionDef => TokenKind::Rpar,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+enum WithItemKind {
+    /// The `with` item is parenthesized, e.g., `with (a, b): ...`.
+    Parenthesized,
+    /// The `with` item has a parenthesized expression, e.g., `with (a) as b: ...`.
+    ParenthesizedExpression,
+    /// The `with` item isn't parenthesized in any way, e.g., `with a as b: ...`.
+    Unparenthesized,
+}
+
+impl WithItemKind {
+    /// Returns the token that terminates a list of `with` items.
+    const fn list_terminator(self) -> TokenKind {
+        match self {
+            WithItemKind::Parenthesized => TokenKind::Rpar,
+            WithItemKind::Unparenthesized | WithItemKind::ParenthesizedExpression => {
+                TokenKind::Colon
+            }
+        }
+    }
+}
+
 #[repr(u8)]
```

Comment:
Nit: I don't think that's still needed, I should have removed it. I experimented with directly casting between kind and the bitflags and that requires a repr but we don't do this anymore.
```suggestion
```